### PR TITLE
Interpreter: Run interpreter test with exchange executors

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:

### What is changed and how it works?
Mock ExchangeSender/ReceiverInputStreams for interpreter tests. Only use mock data, don't care about network and region releted info.
After this pr, one can use mockExecutor to mock all DAGRequest to run interpreter and generate BlockInputStreams.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/DataStreams/MockExchangeSenderInputStream.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    Block res = children.back()->getHeader();
+    return res;
+}
+
+Block MockExchangeSenderInputStream::readImpl()
+{
+    Block res = children.back()->read();
+    if (!res)
+        return res;
+    return res;
```
Comment (SeaRise) [inline dbms/src/DataStreams/MockExchangeSenderInputStream.cpp:49]:
```suggestion
    return children.back()->read();
```

File: dbms/src/DataStreams/MockExchangeSenderInputStream.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        totals = child->getTotals();
+    }
+
+    return totals;
+}
+
+Block MockExchangeSenderInputStream::getHeader() const
+{
+    Block res = children.back()->getHeader();
+    return res;
```
Comment (SeaRise) [inline dbms/src/DataStreams/MockExchangeSenderInputStream.cpp:41]:
```suggestion
    return children.back()->getHeader();
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 3]
```diff
@@ -51,6 +52,9 @@
 #include <Storages/Transaction/TiDB.h>
 #include <WindowFunctions/WindowFunctionFactory.h>
 
+#include "DataStreams/MockExchangeSenderInputStream.h"
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:55]:
```suggestion
#include <DataStreams/MockExchangeSenderInputStream.h>
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 4]
```diff
@@ -682,6 +686,21 @@ void DAGQueryBlockInterpreter::handleExchangeReceiver(DAGPipeline & pipeline)
     analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
 }
 
+void DAGQueryBlockInterpreter::handleMockExchangeReceiver(DAGPipeline & pipeline)
+{
+    for (size_t i = 0; i < max_streams; ++i)
+    {
+        pipeline.streams.push_back(std::make_shared<MockExchangeReceiverInputStream>(query_block.source->exchange_receiver(), context.getSettingsRef().max_block_size));
+    }
+    std::vector<NameAndTypePair> source_columns;
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:695]:
```suggestion
    NamesAndTypes source_columns;
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 5]
```diff
@@ -682,6 +686,21 @@ void DAGQueryBlockInterpreter::handleExchangeReceiver(DAGPipeline & pipeline)
...
+    Block block = pipeline.firstStream()->getHeader();
+    for (const auto & col : block.getColumnsWithTypeAndName())
+    {
+        source_columns.emplace_back(NameAndTypePair(col.name, col.type));
+    }
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:700]:
```suggestion
    for (const auto & col : pipeline.firstStream()->getHeader())
    {
        source_columns.emplace_back(col.name, col.type);
    }
```

File: dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        String name = "exchange_receiver_" + std::to_string(i);
+        ColumnInfo info = TiDB::fieldTypeToColumnInfo(receiver.field_types(i));
+        schema.emplace_back(std::move(name), std::move(info));
+    }
+    for (auto & dag_col : schema)
+    {
+        auto tp = getDataTypeByColumnInfoForComputingLayer(dag_col.second);
+        ColumnWithTypeAndName col(tp, dag_col.first);
+        columns.emplace_back(col);
+    }
```
Comment (SeaRise) [inline dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp:37]: It seems that we can do without DAGSchema.
```
    for (int i = 0; i < receiver.field_types_size(); ++i)
    {
         columns.emplace_back(
             getDataTypeByColumnInfoForComputingLayer(TiDB::fieldTypeToColumnInfo(receiver.field_types(i))), 
             fmt::format("exchange_receiver_{}", i));
    }
```

File: dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+#include <DataStreams/MockExchangeReceiverInputStream.h>
+#include <Flash/Coprocessor/ChunkCodec.h>
+
+namespace DB
+{
+MockExchangeReceiverInputStream::MockExchangeReceiverInputStream(const tipb::ExchangeReceiver & receiver, size_t max_block_size)
+    : output_index(0)
+    , max_block_size(max_block_size)
+{
+    rows = 0;
```
Comment (SeaRise) [inline dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp:24]: move to L38

File: dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+ColumnPtr MockExchangeReceiverInputStream::makeColumn(ColumnWithTypeAndName elem) const
+{
+    auto column = elem.type->createColumn();
+    size_t row_count = 0;
+    for (size_t i = output_index; i < rows & row_count < max_block_size; ++i)
+    {
+        column->insert((*elem.column)[i]);
+        row_count++;
```
Comment (SeaRise) [inline dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp:55]:
```suggestion
        ++row_count;
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 9]
```diff
@@ -51,6 +53,9 @@
 #include <Storages/Transaction/TiDB.h>
 #include <WindowFunctions/WindowFunctionFactory.h>
 
+#include "Core/NamesAndTypes.h"
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:56]:
```suggestion
#include <Core/NamesAndTypes.h>
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 10]
```diff
@@ -919,4 +949,4 @@ BlockInputStreams DAGQueryBlockInterpreter::execute()
 
     return pipeline.streams;
 }
-} // namespace DB
+} // namespace DB
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:952]:
```suggestion
} // namespace DB

```

File: dbms/src/Flash/tests/gtest_interpreter.cpp [hunk 11]
```diff
@@ -249,9 +252,96 @@ CreatingSets
...
+     Expression
+      MockExchangeReceiver)";
+        ASSERT_BLOCKINPUTSTREAM_EQAUL(expected, request, 10);
+    }
 }
 CATCH
 
-
 } // namespace tests
 } // namespace DB
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_interpreter.cpp:347]:
```suggestion
} // namespace DB

```

File: dbms/src/Flash/tests/gtest_interpreter.cpp [hunk 12]
```diff
@@ -249,9 +252,96 @@ CreatingSets
...
+       Expression
+        Expression
+         Expression
+          Expression
+           Expression
+            MockExchangeReceiver)";
+        ASSERT_BLOCKINPUTSTREAM_EQAUL(expected, request, 10);
+    }
+
+    // only join + ExchangeReceiver
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_interpreter.cpp:302]: add case `join + receiver + sender`

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 13]
```diff
@@ -850,7 +870,10 @@ void DAGQueryBlockInterpreter::executeImpl(DAGPipeline & pipeline)
     // execute exchange_sender
     if (query_block.exchange_sender)
     {
-        handleExchangeSender(pipeline);
+        if (dagContext().isTest())
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:873]: Do we need `if (unlikely(dagContext().isTest()))`?
 What do you think about it?

**Comment by SeaRise** [comment]
other LGTM

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp (hunk 13 above)
Comment (ywqzzy) [reply dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:873]: Good idea, I will fix them all.

**Comment by ywqzzy** [comment]
The latest commit can not pass the test, I pushed it by accident.

**Review by SeaRise — APPROVED** [review]

File: dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    , max_block_size(max_block_size)
+{
+    for (int i = 0; i < receiver.field_types_size(); ++i)
+    {
+        columns.emplace_back(
+            getDataTypeByColumnInfoForComputingLayer(TiDB::fieldTypeToColumnInfo(receiver.field_types(i))),
+            fmt::format("exchange_receiver_{}", i));
+    }
+    rows = 0;
+    for (const auto & elem : columns)
```
Comment (Lloyd-Pottiger) [inline dbms/src/DataStreams/MockExchangeReceiverInputStream.cpp:31]: How about iterator in reverse: `for (const auto & elem : reverse(columns))`

**Review by yibin87 — APPROVED** [review]
LGTM

**Comment by yibin87** [comment]
please help to change the MockTableScanBlockInputStream::makeColumn function, the & operator in for statement, which will cause compilation error for gcc. @ywqzzy

**Pull request merged**